app/services/npq/application/accept.rb

```diff
@@ -106,8 +106,14 @@ def participant_profile
         @participant_profile ||= ParticipantProfile.find(npq_application.id)
       end
 
+      def trn
+        @trn ||= npq_application.teacher_reference_number_verified? ? npq_application.teacher_reference_number : user.teacher_profile&.trn
+      end
+
+      # Make sure existing or the new teacher profile has a valid TRN so we fetch same_trn_user & other_accepted_applications_with_same_course correctly
       def teacher_profile
-        @teacher_profile ||= user.teacher_profile || user.build_teacher_profile
+        @teacher_profile ||= (user.teacher_profile&.assign_attributes(trn:)
```

Comment:
I wonder if instead of assigning attributes here, we use the new trn method you've added as follows in the query, so it's directly looking up the trn instead:

```
 def same_trn_user
  return if trn.blank?

  @same_trn_user ||= TeacherProfile
                     .oldest_first
                     .where(trn: trn)
                     .where.not(id: teacher_profile.id)
                     .first
                     &.user
end
```
Let me know your thoughts :)